core/lib/zksync_core/src/reorg_detector/mod.rs

```diff
@@ -231,122 +238,147 @@ impl ReorgDetector {
         &self.health_check
     }
 
+    /// Returns `Ok(())` if no reorg was detected.
+    /// Returns `Err::ReorgDetected()` if a reorg was detected.
+    pub async fn check_consistency(&mut self) -> Result<(), Error> {
+        let mut storage = self
+            .pool
+            .access_storage()
+            .await
+            .context("access_storage()")?;
+        let Some(local_l1_batch) = storage
+            .blocks_dal()
+            .get_last_l1_batch_number_with_metadata()
+            .await
+            .context("get_last_l1_batch_number_with_metadata()")?
+        else {
+            return Ok(());
+        };
+        let Some(local_miniblock) = storage
+            .blocks_dal()
+            .get_sealed_miniblock_number()
+            .await
+            .context("get_sealed_miniblock_number()")?
+        else {
+            return Ok(());
+        };
+        drop(storage);
+
+        let remote_l1_batch = self.client.sealed_l1_batch_number().await?;
+        let remote_miniblock = self.client.sealed_miniblock_number().await?;
+
+        let checked_l1_batch = local_l1_batch.min(remote_l1_batch);
+        let checked_miniblock = local_miniblock.min(remote_miniblock);
+
+        let root_hashes_match = self.root_hashes_match(checked_l1_batch).await?;
+        let miniblock_hashes_match = self.miniblock_hashes_match(checked_miniblock).await?;
+
+        // The only event that triggers re-org detection and node rollback is if the
+        // hash mismatch at the same block height is detected, be it miniblocks or batches.
+        //
+        // In other cases either there is only a height mismatch which means that one of
+        // the nodes needs to do catching up; however, it is not certain that there is actually
+        // a re-org taking place.
+        if root_hashes_match && miniblock_hashes_match {
+            self.event_handler
+                .update_correct_block(checked_miniblock, checked_l1_batch);
+            return Ok(());
+        }
+        let diverged_l1_batch = checked_l1_batch + (root_hashes_match as u32);
+        self.event_handler.report_divergence(diverged_l1_batch);
+
+        // Check that the first L1 batch matches, to make sure that
+        // we are actually tracking the same chain as the main node.
+        let mut storage = self
+            .pool
+            .access_storage()
+            .await
+            .context("access_storage()")?;
+        let first_l1_batch = storage
+            .blocks_dal()
+            .get_earliest_l1_batch_number_with_metadata()
+            .await
+            .context("get_earliest_l1_batch_number_with_metadata")?
+            .context("all L1 blocks dissapeared")?;
```

Comment:
Nit: "all L1 batches"; "L1 blocks" can be interpreted as blocks of the L1 blockchain.

Reply:
done